codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,297 @@
+            assert iNextState == null || iNextState.pseudoTime >= iState.pseudoTime : "iNextState.pseudoTime(" +
+                    iNextState.pseudoTime + ") < " + " iState.pseudoTime(" + iState.pseudoTime + ")";
+            int nsent = distribute(iNextState == null ? maxBytes :
+                                    min(maxBytes, (int) ((iNextState.pseudoTime - iState.pseudoTime) *
+                                             ((double) iNextState.stream.weight() / oldTotalQueuedWeights)
+                                             + allocationQuantum)),
+                               writer,
+                               iState);
+            iState.updatePseudoTime(nsent, oldTotalQueuedWeights);
+            return nsent;
+        } finally { // Do in finally to ensure the internal state is not corrupted if an exception is thrown.
+            // The offer operation is delayed until we unroll up the recursive stack, so we don't have to remove from
+            // the priority queue due to a write operation.
+            if (iState.isActiveCountForTree != 0) {
+                pState.offer(iState);
+            }
+        }
+    }
+
+    private State state(Http2Stream stream) {
+        return stream.getProperty(stateKey);
+    }
+
+    /**
+     * For testing only!
+     */
+    int streamableBytes(Http2Stream stream) {
+        return state(stream).streamableBytes;
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class State implements PriorityQueueNode<State> {
+        final Http2Stream stream;
+        private final Queue<State> queue;
+        int streamableBytes;
+        int isActiveCountForTree;
+        private int priorityQueueIndex = INDEX_NOT_IN_QUEUE;
+        long pseudoTime;
+        long totalQueuedWeights;
+        boolean isActive;
+
+        State(Http2Stream stream) {
+            this(stream, 0, connectionState.pseudoTime);
+        }
+
+        State(Http2Stream stream, int initialSize, long pseudoTime) {
+            this.stream = stream;
+            queue = new PriorityQueue<State>(initialSize);
+            this.pseudoTime = pseudoTime;
+        }
+
+        void write(int numBytes, Writer writer) throws Http2Exception {
+            try {
+                writer.write(stream, numBytes);
+            } catch (Throwable t) {
+                throw connectionError(INTERNAL_ERROR, t, "byte distribution write error");
+            }
+        }
+
+        void isActiveCountChangeForTree(int increment) {
+            assert isActiveCountForTree + increment >= 0;
+            isActiveCountForTree += increment;
+            if (!stream.isRoot()) {
+                State pState = state(stream.parent());
+                if (isActiveCountForTree == 0) {
+                    pState.remove(this);
+                } else if (isActiveCountForTree - increment == 0) { // if frame count was 0 but is now not, then queue.
+                    pState.offer(this);
```

Comment:
There will need to be an initialization of `psuedoTime` equal to the parent's current time just before this queuing.